Readme and code (#1)
friedcucumber/maui-role-based-data-access: README.md

# Role-Based Data Access with the DevExpress Web API Service

This example uses our free [.NET App Security Library & Web API Service](https://www.devexpress.com/products/net/application_framework/security-web-api-service.xml) to implement authentication and role-based data access. We ran a built-in wizard to generate a ready-to-use authentication service. This service uses [Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/) to access a database. A .NET MAUI application sends requests to the Web API Service to obtain or modify data.

<img src="https://user-images.githubusercontent.com/12169834/228174231-0d1f2d88-8b2b-4db4-8969-fd9b2379d8ce.png" width="30%"/>

> If you are new to the DevExpress Web API Service Library, you may want to review the following resources. 

Reviewer: ```suggestion
If you are new to the DevExpress Web API Service Library, you may want to review the following resources:
```

[Create a Standalone Web API Application](https://docs.devexpress.com/eXpressAppFramework/403401/backend-web-api-service/create-new-application-with-web-api-service?p=net6)

[A 1-Click Solution for CRUD Web API with Role-based Access Control via EF Core & ASP.NET](https://www.youtube.com/watch?v=T7y4gwc1n4w&list=PL8h4jt35t1wiM1IOux04-8DiofuMEB33G)

## Prerequisites

[SQL Server](https://www.microsoft.com/en-us/sql-server/sql-server-downloads), if you run this solution on Windows.

## Run Projects

1. Run Visual Studio as Administrator and open the solution. Administrator privileges allow the IDE to create a database when you run the Web Service project. 

> 2. Select **WebApi** in the **debug** dropdown menu. This choice enables [Kestrel](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-7.0) as the web server for debug runs.

Reviewer: ```suggestion
2. Select **WebApi** in the **debug** drop-down menu. This choice enables [Kestrel](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-7.0) as the web server for debug runs.
```

    ![Run Settings](images/authenticate-run-settings.png)

>     If you prefer IIS Express to Kestrel, select **IIS Express** from the **debug** drop-down menu, and use an external text editor to add the following code to `.vs\MAUI_WebAPI\config\applicationhost.config`:

Author: ```suggestion
    If you prefer IIS Express to Kestrel, select **IIS Express** in the **debug** drop-down menu. Use an external text editor to add the following code to `.vs\MAUI_WebAPI\config\applicationhost.config`:
```

            return await HttpClient.PostAsync($"{ApiUrl}Authentication/Authenticate",
                                                new StringContent(JsonSerializer.Serialize(new { userName, password = $"{password}" }), Encoding.UTF8,
                                                ApplicationJson));
      }
    ```

    Include the token in [HttpClient.DefaultRequestHeaders.Authorization](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httprequestheaders.authorization?view=net-7.0). All subsequent requests can then access private endpoints and data: 

    ```csharp
    HttpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", await tokenResponse.Content.ReadAsStringAsync());
    ```

  File to Look At: [WebAPIService.cs](CS/MAUI/Services/WebAPIService.cs)

* WebApi service contains the following custom endpoints:

    * The **CanDeletePost** endpoint allows you to send a request from a mobile device to the service and check whether the current user can delete posts. This allows you to show/hide the delete button in the UI.

        File to Look At: [Updater.cs](CS/WebAPI/Controllers/CustomEndpointController.cs)

>     * The **CurrentUser** endpoint retrieves the current user. You can use it to get the avatar of the logged-in user.

Reviewer 2: ```suggestion
    * The **CurrentUser** endpoint returns information about the authenticated user.
```

        File to Look At: [Updater.cs](CS/WebAPI/Controllers/CustomEndpointController.cs)

>     * The **GetAuthorImage** endpoint retrieves an image by a user ID. A user image does not come with the user object to allow you to load them separately.

Reviewer 2: ```suggestion
    * The **GetAuthorImage** endpoint retrieves an author image by user ID. 
```

        File to Look At: [Updater.cs](CS/WebAPI/Controllers/PublicEndpointController.cs)

>     * The **GetAuthorImage** endpoint retrieves a post image by a post ID. A post image does not come with the user object to allow you to load them separately.

Reviewer 2: The name is not right. I assume it's `GetPostImage`, but I may be wrong. 

```suggestion
    * The **GetPostImage** endpoint retrieves an image by post ID. 
```

        File to Look At: [Updater.cs](CS/WebAPI/Controllers/PublicEndpointController.cs)

> * To create users and specify their passwords, use the `Updater.UpdateDatabaseAfterUpdateSchema` method. You can modify a user's password directly in the database or use the full version of our [XAF UI](https://docs.devexpress.com/eXpressAppFramework/112649/data-security-and-safety/security-system/authentication/passwords-in-the-security-system).

Reviewer 2: I don't understand why it should be XAF UI. Why can't I use any custom application to access that database? I can change it to say that "XAF is an easy option". But I don't see why it should be "the only option". 

```suggestion
* The `Updater.UpdateDatabaseAfterUpdateSchema` method generates users and specifies their login credentials. You can modify a user's password directly in the database. Note: Our cross-platform .NET Application Framework - [XAF UI](https://docs.devexpress.com/eXpressAppFramework/112649/data-security-and-safety/security-system/authentication/passwords-in-the-security-system) - allows you to quickly build a desktop or web UI that accesses the same database.
```

Author: ```suggestion
* The `Updater.UpdateDatabaseAfterUpdateSchema` method generates users and specifies their login credentials. You can modify a user's password directly in the database. Note: Our cross-platform .NET Application Framework - [XAF UI](https://docs.devexpress.com/eXpressAppFramework/112649/data-security-and-safety/security-system/authentication/passwords-in-the-security-system) allows you to quickly build a desktop or web UI that accesses the same database.
```

Author: @Abadzhev, can we mention the XAF UI in the parenthesis and get rid of double dashes (as in my suggestion above)?

    File to Look At: [Updater.cs](CS/WebAPI/DatabaseUpdate/Updater.cs)

> * Use the `PermissionPolicyRole` objects in the `Updater` class to add a user permissions. The following code snippet calls the `AddObjectPermissionFromLambda` method to configure the "Viewer" role that allows the user to read published posts:

Reviewer 2: ```suggestion
* `PermissionPolicyRole` objects in the `Updater` class add user permissions. The following code snippet calls the `AddObjectPermissionFromLambda` method to configure the "Viewer" role (allow the user to read published posts):
```

    ```csharp
    role.AddObjectPermissionFromLambda(SecurityOperations.Read, p => p.IsPublished, SecurityPermissionState.Allow);
    ```

    File to Look At: [Updater.cs](CS/WebAPI/DatabaseUpdate/Updater.cs)

> * The `AddTypePermissionsRecursively` method adds CRUD (create, read, update, delete) permissions for the `Post` type to the "Editor" users (Alex, Antony, and Dennis):

Reviewer 2: "for the Post type to the Editor users" - this phrase is a little confusing, so I decided to restructure the text a bit. 

```suggestion
* The `AddTypePermissionsRecursively` method modifies privileges for the "Editor" role (Alex, Antony, and Dennis). The method adds CRUD permissions (create, read, update, delete) for the `Post` type:
```

    
    ```csharp
    role.AddTypePermissionsRecursively<Post>(SecurityOperations.Read | SecurityOperations.Write | SecurityOperations.Create | SecurityOperations.DeleteObject, SecurityPermissionState.Allow);
    ```

    File to Look At: [Updater.cs](CS/WebAPI/DatabaseUpdate/Updater.cs)

### Login UI and View Model

> * Use the [TextEdit.StartIcon](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.StartIcon) and [PasswordEdit.StartIcon](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.StartIcon) properties to display icons in the [TextEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.TextEdit) and [PasswordEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.PasswordEdit) controls.

Reviewer 2: ```suggestion
* Use [TextEdit.StartIcon](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.StartIcon) and [PasswordEdit.StartIcon](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.StartIcon) properties to display icons in [TextEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.TextEdit) and [PasswordEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.PasswordEdit) controls.
```

Author: Can you describe please, why have you removed these articles?

    ```xaml
    <dxe:TextEdit LabelText="Login" StartIcon="editorsname" .../>
    <dxe:PasswordEdit LabelText="Password" StartIcon="editorspassword" .../>
    ```

    File to Look At: [LoginPage.xaml](CS/MAUI/Views/LoginPage.xaml)
    

> * To validate the [PasswordEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.PasswordEdit) control's value, use the [EditBase.HasError](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.HasError) and [EditBase.ErrorText](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.ErrorText) inherited properties.

Reviewer 2: ```suggestion
* To validate user input in the [PasswordEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.PasswordEdit) control, use [EditBase.HasError](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.HasError) and [EditBase.ErrorText](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.ErrorText) inherited properties.
```

        // ...
        string errorText;
        bool hasError;
        // ...

        public string ErrorText {
            get => errorText;
            set => SetProperty(ref errorText, value);
        }

        public bool HasError {
            get => hasError;
            set => SetProperty(ref hasError, value);
        }
        async void OnLoginClicked() {
            /// ...
            string response = await DataStore.Authenticate(userName, password);
            if (!string.IsNullOrEmpty(response)) {
                ErrorText = response;
                HasError = true;
                return;
            }
            HasError = false;
            await Navigation.NavigateToAsync<SuccessViewModel>();
        }
    }
    ```

    File to Look At: [LoginViewModel.cs](CS/MAUI/ViewModels/LoginViewModel.cs)

* Specify the [TextEdit.ReturnType](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.ReturnType) inherited property to focus the [PasswordEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.PasswordEdit) control after the [TextEdit](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.TextEdit) control's value is edited.

> * Bind the [PasswordEdit.ReturnCommand](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.ReturnCommand) property to the **Login** command to execute the command when a user enters the password:

Reviewer 2: One less mention of the word "command". 

```suggestion
* Use the [PasswordEdit.ReturnCommand](https://docs.devexpress.com/MAUI/DevExpress.Maui.Editors.EditBase.ReturnCommand) property to specify a command (**Login**) that runs when a user enters the password:
```

    ```xaml
    <dxe:PasswordEdit ReturnCommand="{Binding LoginCommand}"/>
    ```

    File to Look At: [LoginPage.xaml](CS/MAUI/Views/LoginPage.xaml)
    
    ```csharp
    public class LoginViewModel : BaseViewModel {
        // ...
        public LoginViewModel() {
            LoginCommand = new Command(OnLoginClicked);
            SignUpCommand = new Command(OnSignUpClicked);
            PropertyChanged +=
                (_, __) => LoginCommand.ChangeCanExecute();

        }
        // ...
        public Command LoginCommand { get; }
        public Command SignUpCommand { get; }
        // ...
    }
    ```

    File to Look At: [LoginViewModel.cs](CS/MAUI/ViewModels/LoginViewModel.cs)

> * In the UI, the Image object caches its displayed image.  because we pass a string with a Uri to the Image. Image caching is described in the [MAUI documentation](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/image?view=net-maui-7.0#load-a-remote-image). To create a Uri, we use a [MultiBinding](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/multibinding?view=net-maui-7.0) that gets the host name and the author or post ID:

Reviewer 2: ```suggestion
* We enabled image caching in this project. To achieve that, we needed to identify images by their Uri. To create a Uri, we use a [MultiBinding](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/multibinding?view=net-maui-7.0) that obtains the host name and the author/post ID. For additional information on image caching, refer to [MAUI documentation](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/image?view=net-maui-7.0#load-a-remote-image). 
```
